[WNMGDS-1466] icon title updates

scheul93 opened the pull request:
## Summary

* Updated the `SvgIcon` to have a `hideTitle` prop and give consuming teams the ability to not render the title prop if they have another accessible label.
* Updated the `TooltipIcon` component to have a screen-reader accessible, optional label other than the title, so that it doesn't conflict with tooltip content
* Updated the tooltip icon tests to use testing library

## How to test

`yarn storybook`. Navigate to Components > Tooltip. Hover over the tooltip icon and confirm that the small grey 'title' no longer appears

## Outstanding questions
* Making this change to SvgIcon changed some of our internal components that are using icons. I _think_ most of them are okay being hidden to screen readers, but I need a second opinion

scheul93 commented on packages/design-system/src/components/UsaBanner/__snapshots__/UsaBanner.test.jsx.snap:
```diff
@@ -11,20 +11,13 @@ exports[`UsaBanner renders correctly 1`] = `
     >
       <svg
```
The USA Banner component is where I'm unsure how the icons should behave

ericwbailey replied:
My thinking here:

- The US flag is visually significant, so it should have a title.
- The down chevron is decorative and should not have a title, as the accessible name and ARIA describe functionality
- The circle icons are decorative, as the subheadings and copy describe the functionality
- The lock icon is visually significant, so it should have a title, and that title should be consistent across the whole site where the lock icon is used in this context.

scheul93 replied:
Thanks for your input! I updated accordingly

pwolfert commented on packages/design-system/src/components/Accordion/AccordionItem.tsx:
```diff
@@ -99,9 +99,13 @@ export class AccordionItem extends React.Component<AccordionItemProps, Accordion
               {isItemOpen ? (
-                <RemoveIcon className="ds-c-accordion__button-icon" title="Close" />
+                <RemoveIcon
+                  className="ds-c-accordion__button-icon"
+                  title="Close"
+                  ariaHidden={false}
+                />
               ) : (
-                <AddIcon className="ds-c-accordion__button-icon" title="Open" />
+                <AddIcon className="ds-c-accordion__button-icon" title="Open" ariaHidden={false} />
```
Is there a way we could do this where we don't have to explicitly set `ariaHidden` to `false`? Shouldn't showing the title be the default behavior?

pwolfert replied:
Actually, it looks like you don't need to define this here

scheul93 replied:
ariaHidden doesn't just set the `title` to not render, it also tells a screenreader that the icon is decorative, which is what _most_ of our icons seem to be. When I implemented the SvgIcon it was recommended that default behavior is that aria-hidden should be true.

I'm not sure what 
> Actually, it looks like you don't need to define this here

is referring to

pwolfert replied:
Okay, I understand now. It's not about hiding the title or not hiding the title here but a new consideration about how this element should behave with screen readers?

pwolfert replied:
That is, it's not directly related to the title business?

ericwbailey commented on packages/design-system/src/components/Dialog/__snapshots__/Dialog.test.jsx.snap:
```diff
@@ -259,19 +253,13 @@ exports[`Dialog renders react-aria-modal 1`] = `
                               className="ds-c-icon ds-c-icon--close "
                               focusable={false}
                               id="icon-1"
                               role="img"
```
```suggestion
```

ericwbailey commented on packages/design-system/src/components/Dialog/__snapshots__/Dialog.test.jsx.snap:
```diff
@@ -131,19 +131,13 @@ exports[`Dialog renders react-aria-modal 1`] = `
                         class="ds-c-icon ds-c-icon--close "
                         focusable="false"
                         id="icon-1"
                         role="img"
```
```suggestion
```

I think we should remove `role="img"` when `aria-hidden` is declared, so as to avoid a role accidentally getting announced.

ericwbailey commented on packages/design-system/src/components/Tooltip/__snapshots__/Tooltip.test.jsx.snap:
```diff
@@ -12,19 +12,13 @@ exports[`Tooltip renders custom trigger component 1`] = `
       class="ds-c-icon ds-c-icon--info-circle-thin ds-c-tooltip-icon"
       focusable="false"
       id="icon-6"
       role="img"
```
```suggestion
```

ericwbailey commented on packages/design-system/src/components/Tooltip/__snapshots__/Tooltip.test.jsx.snap:
```diff
@@ -46,19 +40,13 @@ exports[`Tooltip renders default trigger icon 1`] = `
       class="ds-c-icon ds-c-icon--info-circle-thin ds-c-tooltip-icon"
       focusable="false"
       id="icon-2"
       role="img"
```
```suggestion
```

ericwbailey commented on packages/design-system/src/components/Tooltip/__snapshots__/Tooltip.test.jsx.snap:
```diff
@@ -106,19 +94,13 @@ exports[`Tooltip renders inverse tooltip 1`] = `
           class="ds-c-icon ds-c-icon--info-circle-thin ds-c-tooltip-icon"
           focusable="false"
           id="icon-4"
           role="img"
```
```suggestion
```

ericwbailey commented on packages/design-system/src/components/Tooltip/__snapshots__/Tooltip.test.jsx.snap:
```diff
@@ -207,19 +189,13 @@ exports[`Tooltip tooltip with close renders heading element and close button 1`]
           class="ds-c-icon ds-c-icon--close ds-c-icon--close-thin "
           focusable="false"
           id="icon-18"
           role="img"
```
```suggestion
```

ericwbailey commented on packages/design-system/src/components/Tooltip/__snapshots__/TooltipIcon.test.jsx.snap:
```diff
@@ -1,21 +1,47 @@
+      class="ds-c-icon ds-c-icon--info-circle-thin ds-c-tooltip-icon ds-c-tooltip-icon--inverse"
+      focusable="false"
+      id="icon-2"
+      role="img"
```
```suggestion
```

ericwbailey commented on packages/design-system/src/components/Tooltip/__snapshots__/TooltipIcon.test.jsx.snap:
```diff
@@ -1,21 +1,47 @@
+      class="ds-c-icon ds-c-icon--info-circle-thin ds-c-tooltip-icon"
+      focusable="false"
+      id="icon-1"
+      role="img"
```
```suggestion
```

ericwbailey commented on packages/design-system/src/components/UsaBanner/__snapshots__/UsaBanner.test.jsx.snap:
```diff
@@ -11,20 +11,13 @@ exports[`UsaBanner renders correctly 1`] = `
         class="ds-c-icon ds-c-icon--usa-flag ds-c-usa-banner__header-flag"
         focusable="false"
         id="icon-2"
         role="img"
```
```suggestion
```

ericwbailey commented on packages/design-system/src/components/UsaBanner/__snapshots__/UsaBanner.test.jsx.snap:
```diff
@@ -43,20 +36,13 @@ exports[`UsaBanner renders correctly 1`] = `
             class="ds-c-icon ds-c-icon--arrow ds-c-icon--arrow-down ds-c-usa-banner__action-icon"
             focusable="false"
             id="icon-3"
             role="img"
```
```suggestion
```

ericwbailey commented on packages/design-system/src/components/UsaBanner/__snapshots__/UsaBanner.test.jsx.snap:
```diff
@@ -73,20 +59,13 @@ exports[`UsaBanner renders correctly 1`] = `
             class="ds-c-icon ds-c-icon--building-circle ds-c-usa-banner__icon ds-c-icon-color--primary"
             focusable="false"
             id="icon-4"
             role="img"
```
```suggestion
```

ericwbailey commented on packages/design-system/src/components/UsaBanner/__snapshots__/UsaBanner.test.jsx.snap:
```diff
@@ -108,20 +87,13 @@ exports[`UsaBanner renders correctly 1`] = `
             class="ds-c-icon ds-c-icon--lock-circle ds-c-usa-banner__icon"
             focusable="false"
             id="icon-5"
             role="img"
```
```suggestion
```

ericwbailey commented on packages/design-system/src/components/UsaBanner/__snapshots__/UsaBanner.test.jsx.snap:
```diff
@@ -139,20 +111,13 @@ exports[`UsaBanner renders correctly 1`] = `
               class="ds-c-icon ds-c-icon--lock ds-c-usa-banner__lock-image"
               focusable="false"
               id="icon-6"
               role="img"
```
```suggestion
```

ericwbailey commented on packages/ds-healthcare-gov/src/components/Header/__snapshots__/ActionMenu.test.jsx.snap:
```diff
@@ -93,19 +93,13 @@ exports[`ActionMenu logged-in version renders logged-in version 1`] = `
           class="ds-c-icon ds-u-margin-right--1 ds-c-icon--menu"
           focusable="false"
           id="icon-1"
           role="img"
```
```suggestion
```

ericwbailey commented on packages/ds-healthcare-gov/src/components/Header/__snapshots__/ActionMenu.test.jsx.snap:
```diff
@@ -140,19 +134,13 @@ exports[`ActionMenu logged-in version set aria-expanded to true 1`] = `
           class="ds-c-icon ds-c-icon--close ds-c-icon--close-thin ds-u-margin-right--1"
           focusable="false"
           id="icon-2"
           role="img"
```
```suggestion
```

ericwbailey commented on packages/ds-healthcare-gov/src/components/Header/__snapshots__/ActionMenu.test.jsx.snap:
```diff
@@ -199,19 +187,13 @@ exports[`ActionMenu logged-out version renders logged-out version 1`] = `
           class="ds-c-icon ds-u-margin-right--1 ds-c-icon--menu"
           focusable="false"
           id="icon-4"
           role="img"
```
```suggestion
```

ericwbailey reviewed (changes requested):
I can confirm the expected behavior on Storybook, where the SVG icon's presence is nulled when placed inside a tooltip icon button!

The one small tweak I'm suggesting is to remove `role="img"` from SVG where `aria-hidden="true"` has been declared. It's one extra layer of communicating to AT that this isn't an image.

pwolfert commented on packages/design-system/src/components/Tooltip/TooltipIcon.tsx:
```diff
@@ -1,22 +1,34 @@
+   * Optional way to provide an accessible name to the tooltip icon that is not the <title> element.
+   * Use only if the tooltip icon is not contained in an element that has an accessible name.
+   */
+  ariaLabel?: string;
```
Is this prop a misnomer? It's not actually used as an `aria-label` attribute anywhere, is it? I don't care that much, but I thought I'd flag it as potentially needing a new name now that the implementation is decided on.

pwolfert reviewed (commented):
I'm still pretty fuzzy on the different use cases here that correspond to the different combinations, so I'm having trouble reviewing. I'll seek help understanding through a different medium, though

scheul93 commented:
After a conversation, updated so that
- `SvgIcon` uses the `ariaHidden` prop to determine if the title is shown
- `SvgIcon` updates which attributes it renders on the svg based on `ariaHidden`
- `SvgIcon` conditionally shows `description` now taking into account `ariaHidden`
- `TooltipIcon` no longer includes a `title` or label by default
- Updated `TooltipIcon` guidance to suggest that TooltipIcon has an aria-label when used in `Tooltip`
<img width="745" alt="Screen Shot 2022-04-05 at 3 09 20 PM" src="https://user-images.githubusercontent.com/33579665/161851470-d36ea16f-2d2b-4872-9034-3bac030654d0.png">

pwolfert commented on packages/design-system-docs/src/pages/components/Tooltip/_Tooltip.docs.scss:
```diff
@@ -27,6 +27,8 @@ Style guide: components.tooltip.tooltip
 /*
 `<TooltipIcon>`
 
+When using the `<TooltipIcon>` as the only child of `<Tooltip>`, be sure to provide an `aria-label` to ensure an accessible name for the trigger.
```
```suggestion
When using the `<TooltipIcon>` as the only child of `<Tooltip>`, be sure to provide an `aria-label` on the `<Tooltip>` to ensure an accessible name for the trigger.
```

pwolfert replied:
I like the idea of updating this documentation. I wonder if it would also be helpful to change the example here to actually be a complete tooltip example, and that would give us a chance to show the `aria-label` in use. Like we could make this like the storybook example. We're also going to be ripping this out soon, so it might not be worth it.

pwolfert commented on packages/design-system/src/components/Icons/SvgIcon.tsx:
```diff
@@ -63,20 +62,25 @@ function SvgIcon({
   const titleId = `${iconId}__title`;
   const descriptionId = `${iconId}__desc`;
   const ariaLabelledBy = description ? `${titleId} ${descriptionId}` : titleId;
+  const isSrVisible = !ariaHidden;
```
Nice, this is easier to read

pwolfert reviewed (approved):
LGTM 👍
